Stack/Core/Stack/State/NodeState.cs

```diff
@@ -2255,44 +2316,47 @@ public virtual void ReportEvent(ISystemContext context, IFilterTarget e)
             bool isInverse,
             NodeState target)
         {
-            if (m_notifiers == null)
+            lock (this.m_notifiersLock)
             {
-                m_notifiers = new List<Notifier>();
-            }
+                if (m_notifiers == null)
+                {
+                    m_notifiers = new List<Notifier>();
+                }
 
-            if (NodeId.IsNull(referenceTypeId))
-            {
-                referenceTypeId = ReferenceTypeIds.HasEventSource;
-            }
+                if (NodeId.IsNull(referenceTypeId))
+                {
+                    referenceTypeId = ReferenceTypeIds.HasEventSource;
+                }
 
-            // check for existing reference.
-            Notifier entry = null;
+                // check for existing reference.
+                Notifier entry = null;
 
-            for (int ii = 0; ii < m_notifiers.Count; ii++)
-            {
-                if (Object.ReferenceEquals(m_notifiers[ii].Node, target))
+                for (int ii = 0; ii < m_notifiers.Count; ii++)
                 {
-                    entry = m_notifiers[ii];
-                    break;
+                    if (Object.ReferenceEquals(m_notifiers[ii].Node, target))
+                    {
+                        entry = m_notifiers[ii];
+                        break;
+                    }
                 }
-            }
 
-            // ensure duplicate references are not left over from the model design.
-            if (!NodeId.IsNull(target.NodeId))
-            {
-                RemoveReference(referenceTypeId, isInverse, target.NodeId);
-            }
+                // ensure duplicate references are not left over from the model design.
+                if (!NodeId.IsNull(target.NodeId))
+                {
+                    RemoveReference(referenceTypeId, isInverse, target.NodeId);
+                }
 
-            if (entry == null)
-            {
-                entry = new Notifier();
-                m_notifiers.Add(entry);
-            }
+                if (entry == null)
+                {
+                    entry = new Notifier();
+                    m_notifiers.Add(entry);
```

Comment:
From previous comment, here we would create a new instance, do copying and adding to the new instance. Then the old collection would not mutate.